### PR TITLE
e2e: ginkgo timeline

### PR DESCRIPTION
**Pull request by pohly**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

This ensures that the failure description for a test includes all relevant information (timeouts, failures, log output) in the order in which the events occurred. This is now supported natively by Ginkgo, therefore we can remove the workarounds that we had for this before (like wrapping the fail handler and using a custom skip method).

#### Which issue(s) this PR fixes:

Related-to: https://github.com/onsi/ginkgo/issues/1069

#### Special notes for your reviewer:

TODO:
- [x] the output tests in Kubernetes still need to be adapted
- [x] ~~truncation of the failure message if it is too large~~ - not really required, dropped

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

File: test/e2e/framework/internal/unittests/cleanup/cleanup_test.go
Comment (chendave) [inline test/e2e/framework/internal/unittests/cleanup/cleanup_test.go:146]: okay, this is what I was thinking yesterday!  nice job!

File: test/e2e/framework/test_context.go
Comment (chendave) [inline test/e2e/framework/test_context.go]: my understanding is that as long as the `--report-dir` is specified, the junit file will not be a full report.

File: test/e2e/framework/internal/output/output.go
Comment (chendave) [inline test/e2e/framework/internal/output/output.go:80]: nit: t.Fatalf("expected one test suite, got %d, junit file: %s", len(actual.TestSuites), string(data))

File: test/e2e/framework/internal/junit/junit.go
Comment (chendave) [inline test/e2e/framework/internal/junit/junit.go:3]: Does this file is actually tested anywhere? if not, why not leave it until the UT is possible?

**Review by chendave — COMMENTED** [review]
LGTM, just some questions or nits.

File: hack/ginkgo-e2e.sh
Comment (pohly) [reply hack/ginkgo-e2e.sh:208]: Fixed.

File: test/e2e/framework/test_context.go
Comment (pohly) [reply test/e2e/framework/test_context.go]: Not worth fixing the commit, this comment gets removed later...

File: hack/ginkgo-e2e.sh
Comment (pohly) [reply hack/ginkgo-e2e.sh:209]: We are passing the option to `ginkgo` (= the CLI), therefore `--junit-report` is the right choice.

But perhaps I misunderstand this comment.

File: test/e2e/framework/log.go
Comment (pohly) [reply test/e2e/framework/log.go:44]: I don't understand. What do you want me to change?

File: test/e2e/framework/test_context.go
Comment (pohly) [reply test/e2e/framework/test_context.go]: Correct.

File: test/e2e/framework/internal/output/output.go
Comment (pohly) [reply test/e2e/framework/internal/output/output.go:80]: Added "JUnit content".

File: test/e2e/framework/internal/junit/junit.go
Comment (pohly) [reply test/e2e/framework/internal/junit/junit.go:3]: This is an oversight: output.go was meant to call it instead of using a copy. Fixed.

**Comment by pohly** [comment]
I fixed commits in https://github.com/kubernetes/kubernetes/compare/63dc239c64850e8e8c3568e9861a86b794e894ec..f63ac1dc3320d370d88bdfb1e84ede2c178358b5
and then had to rebase.

File: test/e2e/framework/log.go
Comment (chendave) [reply test/e2e/framework/log.go:44]: you have an global alias defined: `var Fail = ginkgo.Fail` in this file.

File: hack/ginkgo-e2e.sh
Comment (chendave) [reply hack/ginkgo-e2e.sh:209]: I just meant that it's possible to call this script `ginkgo-e2e.sh` with the junit file name specified as `--ginkgo.junit-report=xxx`, now the filename is renamed from `xxx` to `ginkgo_report.xml`.

File: hack/ginkgo-e2e.sh
Comment (chendave) [reply hack/ginkgo-e2e.sh:258]: I confess that don't know much difference between "${@:-}" and "${@}" here, so I checked with the test script, the result shows that there will no empty parameter appended, there is no difference as far as I can see. But yes, I might be wrong.

File: test/e2e/framework/test_context.go
Comment (pohly) [reply test/e2e/framework/test_context.go]: Let me be more precise: the `junit_01.xml` in `--report-dir` will always be the truncated one. The file written via `--junit-report` will be the complete one.

Comment (chendave) [reply test/e2e/framework/test_context.go]: agreed, this is guarded by the checking of `TestContext.ReportDir`.

File: test/e2e/framework/test_context.go
Comment (chendave) [inline test/e2e/framework/test_context.go:540]: those comments seems like dup with https://github.com/kubernetes/kubernetes/pull/113538/files#diff-daa6e9e53080eff01e28fece2f47c94a37d88d98717c54a0e8ab3d488d1aaaa4R25-R28 and could be removed from here, btw, I am fine with it.

File: test/e2e/framework/log.go
Comment (pohly) [reply test/e2e/framework/log.go:44]: That's for backwards compatibility. I think calling `ginkgo.Fail` directly here is clearer.

File: test/e2e/framework/test_context.go
Comment (pohly) [reply test/e2e/framework/test_context.go:540]: Those are two different items that get documented. I think it's better to repeat some information than to assume that the reader remembers a comment that is made elsewhere.

File: hack/ginkgo-e2e.sh
Comment (pohly) [reply hack/ginkgo-e2e.sh:258]: This here shows the difference:
```
$ cat /tmp/callee.sh 
#!/bin/sh

echo Number of arguments: "$#"

$ cat /tmp/caller.sh 
#!/bin/sh

echo 'With "$@":'
/tmp/callee.sh "$@"

echo 'With "${@:-}":'
/tmp/callee.sh "${@:-}"

$ /tmp/caller.sh 
With "$@":
Number of arguments: 0
With "${@:-}":
Number of arguments: 1
```

File: hack/ginkgo-e2e.sh
Comment (pohly) [reply hack/ginkgo-e2e.sh:209]: Passing `--ginkgo.junit-report=xxx` still works because it is passed to e2e.test after the `--ginkgo.junit-report` argument that the CLI itself adds. However, then the default `ginkgo_report.xml` is empty because the CLI doesn't get the individual XML files where it expects them. I think that's reasonable because the user gets what they explicitly asked for.

**Comment by pohly** [comment]
I noticed some deprecation warnings from ginkgo that I missed earlier and added two commits to address that.

File: hack/ginkgo-e2e.sh
Comment (chendave) [reply hack/ginkgo-e2e.sh:257]: luckily this haven't caused any problem so far.

**Review by chendave — APPROVED** [review]
Not sure if you want to bump to 2.6.1 lately, but this all lgtm.

/lgtm

**Comment by pohly** [comment]
Yes, let's use the latest ginkgo and gomega - both updated.

@aojea : when do you want to merge this? I'm still around this week, then back in the new year.

**Comment by pohly** [comment]
/assign @dims 

For deps review.

**Comment by aojea** [comment]
changing reporting things use to have consequences on tools that process those .... I prefer to wait for new year 

Can you get someone to review all the bashism?

**Comment by pohly** [comment]
> changing reporting things use to have consequences on tools that process those .... I prefer to wait for new year

Fair enough.

> Can you get someone to review all the bashism?

@chendave reviewed the script changes. @chendave, can you re-add your LGTM?

**Comment by liggitt** [comment]
lgtm from a dependency perspective, no new dependencies, made our graph a little shallower 🎉 

```
 	"directDependencies": 212,
 	"transitiveDependencies": 234,
 	"totalDependencies": 285,
-	"maxDepthOfDependencies": 26
+	"maxDepthOfDependencies": 23
 }
```

can tag once the gnarly test changes have review/approve :)

**Comment by dims** [comment]
> lgtm from a dependency perspective, no new dependencies, made our graph a little shallower 🎉

LOVE IT!

File: hack/ginkgo-e2e.sh
Comment (aojea) [inline hack/ginkgo-e2e.sh:165]: don't you use `[[ -n "${E2E_REPORT_DIR:-}" ]] ?

File: hack/ginkgo-e2e.sh
Comment (aojea) [inline hack/ginkgo-e2e.sh:198]: `if [[ -n "${report_dir:-}" ]]; then `?

**Comment by aojea** [comment]
@pohly it was easier to review than I expected, 2 comments to the bash script and LGTM, if no voices against we can merge tomorrow and have still 2 days before christmas, WDYT?

File: hack/ginkgo-e2e.sh
Comment (pohly) [reply hack/ginkgo-e2e.sh]: Both are equivalent. Let me add  `-n` if that is more familiar => https://github.com/kubernetes/kubernetes/compare/b7e6ff2dacdb4c1bf49f82cc9d810af5250e4052..cbb6ea5ca462b9de664d0eba9a0495847537408c

**Comment by pohly** [comment]
> if no voices against we can merge tomorrow and have still 2 days before christmas, WDYT?

Works for me.

**Comment by pohly** [comment]
>  it was easier to review than I expected

Not all of my PRs are as large as PR #112923 :sweat_smile:

**Comment by aojea** [comment]
/lgtm
/approve

@dims or @liggitt for pending approvals

Thanks

**Comment by aojea** [comment]
checked 2 jobs, the new report files are about ~50Mb

**Comment by dims** [comment]
/lgtm
/approve

thanks @aojea and @pohly

**Comment by pohly** [comment]
@liggitt: can you complete the approval? 

staging/src/k8s.io/api is missing (go.mod/sum changes).

**Review by BenTheElder — COMMENTED** [review]
submitting very belated comment for the record, while discussing https://kubernetes.slack.com/archives/C7J9RP96G/p1675982155008399 found that I had a pending comment still ...

File: hack/ginkgo-e2e.sh
Comment (pohly) [reply hack/ginkgo-e2e.sh:168]: With the current approach the script is passing additional parameters to the ginkgo CLI and thus needs to know the report dir.

Perhaps it is possible to not use the CLI for it and instead write full reports from inside the `e2e.test` binary  - let me try that.

Comment (pohly) [reply hack/ginkgo-e2e.sh:168]: See https://github.com/kubernetes/kubernetes/pull/115678

**Pull request merged**
